bindings/sundials4py/kinsol/kinsol.cpp

```diff
@@ -0,0 +1,147 @@
+    [](void* kin_mem, std::function<std::remove_pointer_t<FN_TYPE>> fn) \
+    {                                                                   \
+      auto fntable    = get_kinsol_fn_table(kin_mem);                   \
+      fntable->MEMBER = nb::cast(fn);                                   \
+      if (fn) { return NAME(kin_mem, WRAPPER); }                        \
+      else { return NAME(kin_mem, nullptr); }                           \
+    },                                                                  \
+    __VA_ARGS__)
+
+#define BIND_KINSOL_CALLBACK2(NAME, FN_TYPE1, MEMBER1, WRAPPER1, FN_TYPE2, \
+                              MEMBER2, WRAPPER2, ...)                      \
+  m.def(                                                                   \
+    #NAME,                                                                 \
+    [](void* kin_mem, std::function<std::remove_pointer_t<FN_TYPE1>> fn1,  \
+       std::function<std::remove_pointer_t<FN_TYPE2>> fn2)                 \
+    {                                                                      \
+      auto fntable     = get_kinsol_fn_table(kin_mem);                     \
+      fntable->MEMBER1 = nb::cast(fn1);                                    \
+      fntable->MEMBER2 = nb::cast(fn2);                                    \
+      if (fn1) { return NAME(kin_mem, WRAPPER1, WRAPPER2); }               \
+      else { return NAME(kin_mem, nullptr, WRAPPER2); }                    \
+    },                                                                     \
+    __VA_ARGS__)
+
+namespace sundials4py {
+
+void bind_kinsol(nb::module_& m)
+{
+#include "kinsol_generated.hpp"
+
+  nb::class_<KINView>(m, "KINView")
+    .def("get", nb::overload_cast<>(&KINView::get, nb::const_),
+         nb::rv_policy::reference);
+
+  m.def(
+    "KINSetOptions",
+    [](void* kin_mem, const std::string& kinid, const std::string& file_name,
+       int argc, const std::vector<std::string>& args)
+    {
+      std::vector<char*> argv;
+
+      for (const auto& arg : args)
+      {
+        // We need a non-const char*, so we use data() and an explicit cast.
+        // This is safe as long as the underlying std::string is not modified.
+        argv.push_back(const_cast<char*>(arg.data()));
+      }
+
+      return KINSetOptions(kin_mem, kinid.empty() ? nullptr : kinid.c_str(),
+                           file_name.empty() ? nullptr : file_name.c_str(),
+                           argc, argv.data());
+    },
+    nb::arg("kin_mem"), nb::arg("kinid"), nb::arg("file_name"), nb::arg("argc"),
+    nb::arg("args"));
+
+  m.def(
+    "KINCreate",
+    [](SUNContext sunctx)
+    { return std::make_shared<KINView>(KINCreate(sunctx)); },
+    nb::arg("sunctx"), nb::keep_alive<0, 1>());
+
+  m.def("KINInit",
+        [](void* kin_mem, std::function<std::remove_pointer_t<KINSysFn>> sysfn,
+           N_Vector tmpl)
+        {
+          int kin_status = KINInit(kin_mem, kinsol_sysfn_wrapper, tmpl);
```

Comment:
```suggestion
          if (!sysfn) { throw sundials4py::illegal_value("sysfn was null"); }
          int kin_status = KINInit(kin_mem, kinsol_sysfn_wrapper, tmpl);
          if (kin_status != KIN_SUCCESS)
          {
            throw sundials4py::error_returned("Failed to initialize KINSOL");
          }
```
Should we return the error code or maybe pass the value  to `sundials4py::error_returned`? Alternatively, call `KINGetReturnFlagName` and pass the flag name to `sundials4py::error_returned` instead of the numeric value.